geps/gep-1958.md

```diff
@@ -0,0 +1,44 @@
+# GEP-1958: TLS ALPN Routing
+Issue: #1958
+Status: Provisional
+## TLDR
+Add support for routing traffic based on the application protocols present in the "client hello" step of a TLS handshake.
+
+## Goals
+
+- Provide a way to route a TLS connection to a specific backend based on application protocols.
+- Define expected behavior for routing based on application protocols with respect to ordering and ambiguous matches.
+
+## Non-Goals
+- Match or detect protocols outside of the context of the application protocols present in the TLS client hello.
+- Provide routing based on any other TLS extensions.
+
+## Introduction
+ALPN (application-layer protocol negotiation) is a TLS extension which is present in the TLS client hello.
+It allows a client to specify a list of supported protocols when initiating a TLS connection.
+
+Similar to SNI (server name indication), another TLS extension present in the TLS client hello, ALPN can be used
+to route connections to different backends based on what application protocols the client supports.
+
+With this feature service owners may specify a list of application protocols to match and route TLS connections to a specific backend.
```

Comment:
When we are routing it's based on the *client provided list of accepted ALPN*,right?

This may be odd for passthrough. Cliejt may present h2, http1. Route matches "h2" and routes. Backend selects "http1" .

Now our route that matched "h2" ends up serving http1/1

Reply:
This seems fine IMO as long as long as long as its clear how routing decisions are being made.